src/clojure/nrepl/socket.clj

```diff
@@ -0,0 +1,238 @@
+    (let [make (.getDeclaredMethod junixsocket-server-class "newInstance" nil)]
+      #(.invoke make nil nil))))
+
+(defn unix-server-socket
+  "Returns a filesystem socket bound to the path if the JDK is version
+  16 or newer or if com.kohlschutter.junixsocket/junixsocket-core can
+  be loaded dynamically.  Otherwise throws the ex-info map
+  {:nrepl/kind ::no-filesystem-sockets}."
+  [^String path]
+  (let [^SocketAddress addr (unix-socket-address path)]
+    (case unix-domain-flavor
+      :jdk
+      (let [sock (jdk-unix-server-socket)]
+        (.bind ^ServerSocketChannel sock addr)
+        (let [^Path path (get-path addr)]
+          (-> path .toFile .deleteOnExit))
+        sock)
+
+      :junixsocket
+      (let [sock (junix-server-socket)]
+        (.bind ^ServerSocket sock addr)
+        (let [^String path (get-path addr)]
+          (-> path File. .deleteOnExit))
+        sock)
+
+      (let [msg "Support for filesystem sockets requires JDK 16+ or a junixsocket dependency"]
+        (log msg)
+        (throw (ex-info msg {:nrepl/kind ::no-filesystem-sockets}))))))
+
+(defn as-nrepl-uri [sock transport-scheme]
+  (let [get-local-addr (fn [^NetworkChannel c] (.getLocalAddress c))]
+    (if-let [addr (and (some-> jdk-unix-server-class (instance? sock))
+                       (get-local-addr sock))]
+      (URI. (str transport-scheme "+unix")
+            (let [^Path path (get-path addr)]
+              (-> path .toAbsolutePath str))
+            nil)
+      (if-let [addr (and (some-> junixsocket-server-class (instance? sock))
+                         (.getLocalSocketAddress ^ServerSocket sock))]
+        (URI. (str transport-scheme "+unix")
+              (get-path addr)
+              nil)
+        ;; Assume it's an InetAddress socket
+        (let [sock ^ServerSocket sock]
+          (URI. transport-scheme
+                nil ;; userInfo
+                (-> sock .getInetAddress .getHostName)
+                (.getLocalPort sock)
+                nil       ;; path
+                nil       ;; query
+                nil)))))) ;; fragment
+
+(defprotocol Acceptable
+  (accept [s]
+    "Accepts a connection on s.  Throws ClosedChannelException if s is
+    closed."))
+
+(extend-protocol Acceptable
+  ServerSocketChannel
+  (accept [s] (.accept s))
+
+  ServerSocket
+  (accept [s]
+    (when (.isClosed s)
+      (throw (ClosedChannelException.)))
+    (.accept s)))
+
+;; We have to handle this ourselves for NIO because unfortunately read and write
+;; hang if we use both Channels/newInputStream and Channels/newOutputStream.
+;; Read and write deadlock on a shared channel input/output stream lock
+;; (cf. https://bugs.openjdk.java.net/browse/JDK-4509080).  Verified that this
+;; still happens (via thread dump when hung) with jdk 16.
+
+(definterface Writable
+  (write [byte-array]
+         "Writes the given bytes to the output as per OutputStream write.")
+  (write [byte-array offset length]
```

Comment:
```suggestion
  (write [_byte-array _offset _length]
```

Reply:
Ahh, thanks - I didn't recall that you could use an underscore prefix like that.